homeassistant/components/zha/core/gateway.py

```diff
@@ -231,3 +300,65 @@ def async_update_device(self, sender):
 
         device_entity = async_create_device_entity(zha_device)
         await self._component.async_add_entities([device_entity])
+
+        if is_new_join:
+            device_info = async_get_device_info(self._hass, zha_device)
+            async_dispatcher_send(
+                self._hass,
+                ZHA_GW_MSG,
+                {
+                    TYPE: DEVICE_FULL_INIT,
+                    DEVICE_INFO: device_info
+                }
+            )
+
+
+@callback
+def async_capture_log_levels():
+    """Capture current logger levels for ZHA."""
+    return {
+        BELLOWS: logging.getLogger(BELLOWS).getEffectiveLevel(),
+        ZHA: logging.getLogger(ZHA).getEffectiveLevel(),
+        ZIGPY: logging.getLogger(ZIGPY).getEffectiveLevel(),
+        ZIGPY_XBEE: logging.getLogger(ZIGPY_XBEE).getEffectiveLevel(),
+        ZIGPY_DECONZ: logging.getLogger(ZIGPY_DECONZ).getEffectiveLevel(),
+    }
+
+
+@callback
+def async_set_logger_levels(levels):
+    """Set logger levels for ZHA."""
+    logging.getLogger(BELLOWS).setLevel(levels[BELLOWS])
+    logging.getLogger(ZHA).setLevel(levels[ZHA])
+    logging.getLogger(ZIGPY).setLevel(levels[ZIGPY])
+    logging.getLogger(ZIGPY_XBEE).setLevel(levels[ZIGPY_XBEE])
+    logging.getLogger(ZIGPY_DECONZ).setLevel(levels[ZIGPY_DECONZ])
+
+
+class LogRelayHandler(logging.Handler):
+    """Log handler for error messages."""
+
+    def __init__(self, hass, gateway):
+        """Initialize a new LogErrorHandler."""
+        super().__init__()
+        self.hass = hass
+        self.gateway = gateway
+
+    def emit(self, record):
+        """Relay log message via dispatcher."""
+        if self.gateway.debug_enabled:
+            stack = []
+            if record.levelno >= logging.WARN:
+                if not record.exc_info:
+                    stack = [f for f, _, _, _ in traceback.extract_stack()]
+
+            entry = LogEntry(record, stack,
+                             _figure_out_source(record, stack, self.hass))
+            async_dispatcher_send(
```

Comment:
Is all logging done from async contexts? If not, you will need to check here if your function is inside the loop or not.

Reply:
Can you expand on this a bit? Do you have an example I can peek at? I’m assuming there is some sort of check, and if in loop use this else use regular dispatcher_send? I haven’t seen any issues... but that doesn’t mean this is correct either.

Reply:
Zigpy, bellows, zigpy-deconz and zigpy-xbee all run in event loop. ZHA platform as well. I don't think we have anything sync running.

Reply:
that was my thought... but I can't guarantee everything is perfect.